test/common/common/mutex_tracer_test.cc

```diff
@@ -0,0 +1,102 @@
+#include <chrono>
+#include <thread>
+
+#include "common/common/mutex_tracer.h"
+
+#include "absl/synchronization/mutex.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class MutexTracerTest : public testing::Test {
+protected:
+  void SetUp() {
+    MutexTracer::getOrCreateTracer()->Reset();
+    absl::RegisterMutexTracer(&MutexTracer::contentionHook);
+  }
+
+  absl::Mutex mu_;
+};
+
+// Call the contention hook manually.
+TEST_F(MutexTracerTest, AddN) {
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->numContentions(), 0);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->currentWaitCycles(), 0);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->lifetimeWaitCycles(), 0);
+
+  MutexTracer::getOrCreateTracer()->contentionHook(nullptr, nullptr, 2);
+
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->numContentions(), 1);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->currentWaitCycles(), 2);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->lifetimeWaitCycles(), 2);
+
+  MutexTracer::getOrCreateTracer()->contentionHook(nullptr, nullptr, 3);
+
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->numContentions(), 2);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->currentWaitCycles(), 3);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->lifetimeWaitCycles(), 5);
+
+  MutexTracer::getOrCreateTracer()->contentionHook(nullptr, nullptr, 0);
+
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->numContentions(), 3);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->currentWaitCycles(), 0);
+  EXPECT_EQ(MutexTracer::getOrCreateTracer()->lifetimeWaitCycles(), 5);
+}
+
+void holdMutexForNMilliseconds(absl::Mutex* mu, int64_t duration) {
+  mu->Lock();
+  std::this_thread::sleep_for(std::chrono::milliseconds(duration));
```

Comment:
this will trigger a format-violation, right? Use DangerousDeprecatedTestTime to access RealTimeSystem::sleep().

moreover it looks like it could be pretty flaky. So better yet, use a absl::Synchronization to block here and hold the mutex until the main thread tells you to let it go.